src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc

```diff
@@ -549,11 +554,13 @@ void WeightedRoundRobin::Picker::BuildSchedulerAndStartTimerLocked() {
     gpr_log(GPR_INFO, "[WRR %p picker %p] new weights: %s", wrr_.get(), this,
             absl::StrJoin(weights, " ").c_str());
   }
-  scheduler_ = StaticStrideScheduler::Make(
-      weights,
-      // This requires holding mu_, but we don't have a way to plumb the
-      // annotation through absl::AnyInvocable<>.
-      [this]() ABSL_NO_THREAD_SAFETY_ANALYSIS { return scheduler_state_++; });
+  auto scheduler = StaticStrideScheduler::Make(
+      weights, [this]() { return wrr_->scheduler_state_.fetch_add(1); });
+  if (scheduler.has_value()) {
+    scheduler_ = std::make_shared<StaticStrideScheduler>(std::move(*scheduler));
+  } else {
+    scheduler_.reset();
+  }
```

Comment:
With #31973, since we share the same mu_, it appears the current implementation will block all new pick() calls every time we rebuild the map. If so, this seems to somewhat negate gains from this change and #31973 periodically. Can we have a separate mutex only for scheduler_ e.g. with using a wrapper? As I shared I think pick() is one of the most critical path regarding RPC latency and close optimization is well worth it.

Reply:
Good point.  I've split this up so that we use separate mutexes for the timer and for the scheduler, to minimize the critial period for the latter mutex.